jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentSourceRange.java

```diff
@@ -0,0 +1,187 @@
+     * @param offset the offset byte of the content to start from.
+     *               Must be greater than or equal to 0 and less than the content length (if known).
+     * @param length the length of the content to make available, -1 for the full length.
+     *               If the size of the content is known, the length may be truncated to the content size minus the offset.
+     * @param readToEof - if true, read until EOF of the source after the range is read.
+     * @throws IndexOutOfBoundsException if the offset or length are out of range.
+     * @see TypeUtil#checkOffsetLengthSize(long, long, long)
+     */
+    public ContentSourceRange(Content.Source source, long offset, long length, boolean readToEof)
+    {
+        _source = source;
+        _offset = offset;
+        _length = TypeUtil.checkOffsetLengthSize(offset, length, source.getLength());
+        _offsetRemaining = _offset;
+        _lengthRemaining = _length;
+        _readToEof = readToEof;
+    }
+
+    @Override
+    public long getLength()
+    {
+        return _length;
+    }
+
+    @Override
+    public Content.Chunk read()
+    {
+        while (true)
+        {
+            Content.Chunk chunk = _source.read();
+            if (chunk == null)
+                return null;
+
+            if (Content.Chunk.isFailure(chunk))
+                return chunk;
+
+            if (_offsetRemaining > 0)
+            {
+                if (_offsetRemaining >= chunk.remaining())
+                {
+                    // We can skip this whole chunk.
+                    _offsetRemaining -= chunk.remaining();
+                    if (chunk.isLast())
+                    {
+                        chunk.clear();
+                        return chunk;
+                    }
+
+                    chunk.release();
+                    continue;
+                }
+                else
+                {
+                    // Advance position to the correct offset.
+                    chunk.skip(_offsetRemaining);
+                    _offsetRemaining = 0;
+                }
+            }
+
+            // We can start processing the limited length if we have reached the starting offset.
+            if (_offsetRemaining == 0 && _lengthRemaining >= 0)
+            {
+                if (_lengthRemaining == 0)
+                {
+                    if (!_readToEof)
+                    {
+                        // We do not have to read until EOF of the source, so we can return EOF now and fail the source.
+                        _source.fail(new ClosedChannelException());
+                        return Content.Chunk.EOF;
+                    }
+
+                    // Release the chunk and continue until we find a last chunk.
+                    if (!chunk.isLast())
+                    {
+                        chunk.release();
+                        continue;
+                    }
+
+                    chunk.clear();
```

Comment:
I think again this is missing a `release()`.


```suggestion
                    chunk.release();
                    return Chunk.EOF;
```